File: LAB4/task_4/ElectionSystem.cpp
/**
 * Election.cpp
 * VotingAlgorithms.cpp
 * BordaCondorcet.cpp
 */

/**
 * @file ElectionSystem.cpp
 * @brief Реализация методов Борда и Кондорсе для определения победителей выборов.
 * 
 * Класс Election предоставляет функционал для:
 * - Подсчета голосов по методу Борда (начисление баллов за позиции).
 * - Определения победителя по методу Кондорсе (попарные сравнения).
 * - Обработки крайних случаев: ничья, отсутствие победителя Кондорсе.
 * 
 * Пример использования:
 * @code
 * vector<string> candidates = {"A", "B", "C"};
 * vector<vector<string>> voters = {{"A", "B", "C"}, {"B", "C", "A"}};
 * Election election(candidates, voters);
 * cout << "Борда: " << election.BordaCount() << endl;
 * @endcode
 */

#include <iostream>
#include <vector>
#include <unordered_map>
#include <algorithm>

using namespace std;

class Election {
private:
    vector<string> candidates;
    vector<vector<string>> votersPreferences;
    vector<unordered_map<string, int>> voterRankMaps; ///< Кэш позиций кандидатов для быстрого доступа

    /**
     * @brief Инициализирует кэш позиций кандидатов для каждого избирателя.
     */
    void initRankCache() {
        for (const auto& voter : votersPreferences) {
            unordered_map<string, int> rankMap;
            for (int i = 0; i < voter.size(); ++i) {
                rankMap[voter[i]] = i;
            }
            voterRankMaps.push_back(rankMap);
        }
    }

public:
    /**
     * @brief Конструктор класса Election.
     * @param cand Список кандидатов (должен быть непустым).
     * @param prefs Предпочтения избирателей (каждый подсписок должен содержать всех кандидатов).
     * @throws invalid_argument Если входные данные некорректны.
     */
    Election(vector<string> cand, vector<vector<string>> prefs) 
        : candidates(cand), votersPreferences(prefs) {
        if (candidates.empty() || prefs.empty()) {
            throw invalid_argument("Список кандидатов или избирателей пуст");
        }
        initRankCache();
    }

    /**
     * @brief Вычисляет победителя по методу Борда.
     * @return Имя победителя или "Ничья", если несколько кандидатов имеют максимальный балл.
     */
    string BordaCount() {
        unordered_map<string, int> scores;
        int n = candidates.size();

        for (const auto& voter : votersPreferences) {
            for (size_t i = 0; i < voter.size(); ++i) {
                scores[voter[i]] += (n - 1 - i);
            }
        }

        // Поиск максимального балла
        auto maxIt = max_element(scores.begin(), scores.end(), 
            [](const auto& a, const auto& b) { return a.second < b.second; });

        // Проверка ничьей
        int count = count_if(scores.begin(), scores.end(), 
            [&](const auto& pair) { return pair.second == maxIt->second; });

        return (count == 1) ? maxIt->first : "Ничья";
    }

    /**
     * @brief Вычисляет победителя по методу Кондорсе.
     * @return Имя победителя или "Нет победителя", если такого нет.
     */
    string CondorcetWinner() {
        for (const auto& candidateA : candidates) {
            bool dominatesAll = true;
            for (const auto& candidateB : candidates) {
                if (candidateA == candidateB) continue;

                int aWins = 0;
                for (const auto& rankMap : voterRankMaps) {
                    if (rankMap.at(candidateA) < rankMap.at(candidateB)) {
                        aWins++;
                    }
                }

                if (aWins <= votersPreferences.size() / 2) {
                    dominatesAll = false;
                    break;
                }
            }
            if (dominatesAll) return candidateA;
        }
        return "Нет победителя";
    }
};

File: LAB4/task_4/ElectionSystem.py
class ElectionSystem:
    """
    Класс для определения победителей выборов по методам Борда и Кондорсе.

    Атрибуты:
        candidates (list): Список кандидатов.
        voters_preferences (list): Список предпочтений избирателей.
    """

    def __init__(self, candidates, voters_preferences):
        """
        Инициализация системы выборов.

        Args:
            candidates (list): Список кандидатов.
            voters_preferences (list): Список предпочтений избирателей. 
                Каждый элемент списка — это список кандидатов в порядке убывания предпочтений.
        """
        self.candidates = candidates
        self.voters_preferences = voters_preferences

    def borda_count(self):
        """
        Вычисляет победителя по методу Борда.

        Returns:
            str: Имя кандидата с наибольшим количеством баллов.
        """
        scores = {candidate: 0 for candidate in self.candidates}
        n = len(self.candidates)
        
        for voter in self.voters_preferences:
            for idx, candidate in enumerate(voter):
                scores[candidate] += (n - 1 - idx)
        
        max_score = max(scores.values())
        winners = [k for k, v in scores.items() if v == max_score]
        return winners[0] if len(winners) == 1 else "Ничья"

    def condorcet_winner(self):
        """
        Вычисляет победителя по методу Кондорсе.

        Returns:
            str: Имя кандидата, побеждающего всех в парных сравнениях, или None, если такого нет.
        """
        for candidate_a in self.candidates:
            is_winner = True
            for candidate_b in self.candidates:
                if candidate_a == candidate_b:
                    continue
                a_vs_b = 0
                for voter in self.voters_preferences:
                    if voter.index(candidate_a) < voter.index(candidate_b):
                        a_vs_b += 1
                if a_vs_b <= len(self.voters_preferences) / 2:
                    is_winner = False
                    break
            if is_winner:
                return candidate_a
        return None

    def get_winners(self):
        """
        Возвращает победителей по обоим методам.

        Returns:
            tuple: (победитель Борда, победитель Кондорсе, комментарий)
        """
        borda = self.borda_count()
        condorcet = self.condorcet_winner()
        comment = ""
        if borda != condorcet and condorcet is not None:
            comment = "Разные методы дали разных победителей!"
        return (borda, condorcet, comment)

# Пример использования
if __name__ == "__main__":
    candidates = ["Петя", "Вася", "Оля"]
    voters = [
        ["Вася", "Оля", "Петя"],
        ["Оля", "Петя", "Вася"],
        ["Вася", "Петя", "Оля"]
    ]

    election = ElectionSystem(candidates, voters)
    borda_winner, condorcet_winner, comment = election.get_winners()

    print(f"Победитель по методу Борда: {borda_winner}")
    print(f"Победитель по методу Кондорсе: {condorcet_winner if condorcet_winner else 'Нет победителя'}")
    print(comment)
